src/ios/Utility/THEProtectedMutableDictionary.m

```diff
@@ -0,0 +1,99 @@
+//
+//  The MIT License (MIT)
+//
+//  Copyright (c) 2015 Microsoft
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+//  Thali CordovaPlugin
+//  THEProtectedMutableDictionary.m
+//
+
+#import <pthread.h>
+#import "THEProtectedMutableDictionary.h"
+
+@interface THEProtectedMutableDictionary()
+{
+  NSMutableDictionary * _dict;
+}
+@end
+
+@implementation THEProtectedMutableDictionary
+{
+  pthread_mutex_t _mutex;
+}
+
+- (instancetype)init
+{
+  self = [super init];
+  if (!self)
+  {
+    return nil;
+  }
+ 
+  _dict = [[NSMutableDictionary alloc] init];
+
+  return self;
+}
+
+- (void)updateForKey:(NSObject<NSCopying> *)key updateBlock:(NSObject *(^)(NSObject *))updateBlock
+{
+  // Can't have nil keys
+  if (key == nil)
```

Comment:
NON-BLOCKING - This line really confused me. Not in what it did but in its implications. If you remember at one point I said I wasn't sure how certain code was working. What turned out to be happening is that you had code that called updateForKey with a nil key and was depending on the fact that doing so would mean that essentially the call would be a NOOP. That isn't wrong but it confused me. When someone calls a function I generally expect it to do something, even failures should be 'noisy'. But in this case you were expecting 'something' (calling this function) to do 'nothing' (no call to the update block) and that made reading the code that depending on this 'silent failure' behavior hard to read.